### PR TITLE
O(n) Prime sieve

### PR DESCRIPTION
**Pull request by ongyiumark**
Added O(n) prime sieve. Also, its actually shorter!

Prime sieve seems to be commented out in `notebook.tex`. Should i add it back in?

### CODE REVIEW

File: pvl/abridged/math/numtheory/prime-sieve.cpp [hunk 1]
```diff
@@ -1,12 +1,8 @@
-std::bitset<N> is; // #include <bitset>
-int pr[N], primes = 0;
+bitset<N> isc; // #include <bitset>
+vector<int> p;
```
Comment (leloykun) [inline pvl/abridged/math/numtheory/prime-sieve.cpp:2]: `std::`?

File: pvl/math/prime_sieve.hpp [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+            if (!is_composite[i]) primes.push_back(i);
+            for (int j = 0;
+                j < primes.size() && i*primes[j] < n;
+                j++) {
+                is_composite[i*primes[j]] = 1;
+                if (i%primes[j] == 0) break;
+            }
+        }
+    }
+};
```
Comment (leloykun) [inline pvl/math/prime_sieve.hpp:24]: use 2-space indentation

**Review by leloykun — CHANGES_REQUESTED** [review]

**Comment by leloykun** [comment]
And yah, add it back in

File: pvl/abridged/math/numtheory/prime-sieve.cpp (hunk 1 above)
Comment (ongyiumark) [reply pvl/abridged/math/numtheory/prime-sieve.cpp:2]: oh do we `std::`  in abridged?

File: tests/CMakeLists.txt [hunk 3]
```diff
@@ -48,3 +48,7 @@ gtest_discover_tests(Prim)
 add_executable(ZAlgo ./z.cc)
 target_link_libraries(ZAlgo gtest_main)
 gtest_discover_tests(ZAlgo)
+
+add_executable(PrimeSieve ./prime_sieve.cc)
+target_link_libraries(PrimeSieve gtest_main)
+gtest_discover_tests(PrimeSieve)
```
Comment (leloykun) [inline tests/CMakeLists.txt:54]: - [ ] move `z.cc` to `tests/strings/`
- [ ] move `prime_sieve.cc` to `tests/math/numtheory/`

File: pvl/math/prime_sieve.hpp [hunk 4]
```diff
@@ -0,0 +1,28 @@
+#ifndef PVL_MATH_PRIME_SIEVE_HPP_
+#define PVL_MATH_PRIME_SIEVE_HPP_ 1
```
Comment (leloykun) [inline pvl/math/prime_sieve.hpp:2]: - [ ] move this file to `pvl/math/numtheory/` then fix the guard definition (maybe to `PVL_MATH_NUMTHEORY_SIEVE_HPP_`?)

**Review by leloykun — CHANGES_REQUESTED** [review]

File: pvl/abridged/math/numtheory/prime-sieve.cpp (hunk 1 above)
Comment (leloykun) [reply pvl/abridged/math/numtheory/prime-sieve.cpp:2]: For consistency haha. But, ideally, we shouldn't. I guess it we should start not using it na in abridged from now on

Up to you if you wanna add it; just resolve the thread whenever you want

File: pvl/math/prime_sieve.hpp [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+struct prime_sieve {
+    std::vector<bool> is_composite;
+    std::vector<int> primes;
+    explicit prime_sieve(int n) : is_composite(n, 0) {
+        for (int i = 2; i < n; i++) {
+            if (!is_composite[i]) primes.push_back(i);
+            for (int j = 0;
+                j < primes.size() && i*primes[j] < n;
+                j++) {
```
Comment (leloykun) [inline pvl/math/prime_sieve.hpp:18]: How about this?
```
for (
  int j = 0;
  j < primes.size() && i*primes[j] < n;
  j++
) {
```

File: tests/prime_sieve.cc [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    769, 773, 787, 797, 809, 811, 821, 823, 827, 829, 839,
+    853, 857, 859, 863, 877, 881, 883, 887, 907, 911, 919,
+    929, 937, 941, 947, 953, 967, 971, 977, 983, 991, 997};
+
+    pvl::prime_sieve ps(hiP);
+    int n = ps.primes.size();
+    for (int i = 0; i < n; i++) {
+        EXPECT_EQ(expected[i], ps.primes[i]);
+    }
+}
```
Comment (leloykun) [inline tests/prime_sieve.cc:37]: use 2-space indent here too

(use 2-space indent when coding with C++; 4-space with Python)

**Comment by leloykun** [comment]
One more thing:
- [ ] Add yourself to the contributors section

File: pvl/math/prime_sieve.hpp (hunk 5 above)
Comment (ongyiumark) [reply pvl/math/prime_sieve.hpp:18]: i just renamed the vector to be shorter. It fits in one line now HAHA

File: tests/CMakeLists.txt (hunk 3 above)
Comment (leloykun) [reply tests/CMakeLists.txt:54]: follow `pvl/abridged` for where to put stuff

**Review by leloykun — APPROVED** [review]

**Pull request merged**